src/main/java/com/github/nkzawa/engineio/client/transports/WebSocket.java

```diff
@@ -37,89 +64,126 @@ protected void doOpen() {
+                for (int i = 0; i < response.headers().size(); ++i) {
+                    headers.put(response.headers().name(i), response.headers().value(i));
                 }
-                @Override
-                public void onMessage(final ByteBuffer s) {
-                    EventThread.exec(new Runnable() {
-                        @Override
-                        public void run() {
-                            self.onData(s.array());
-                        }
-                    });
+                EventThread.exec(new Runnable() {
+                    @Override
+                    public void run() {
+                        self.emit(EVENT_RESPONSE_HEADERS, headers);
+                        self.onOpen();
+                    }
+                });
+            }
+
+            @Override
+            public void onMessage(final BufferedSource payload, final com.squareup.okhttp.ws.WebSocket.PayloadType type) throws IOException {
+                logger.fine("onMessage, type: " + type);
+                Object data;
+                switch (type) {
+                    case TEXT:
+                        data = payload.readUtf8();
+                        break;
+                    case BINARY:
+                        data = payload.readByteArray();
+                        break;
+                    default:
+                        throw new IllegalStateException("Unknown payload type: " + type);
                 }
-                @Override
-                public void onError(final Exception e) {
-                    EventThread.exec(new Runnable() {
-                        @Override
-                        public void run() {
-                            self.onError("websocket error", e);
+                payload.close();
+                final Object finalData = data;
+                EventThread.exec(new Runnable() {
+                    @Override
+                    public void run() {
+                        if (finalData instanceof String) {
+                            self.onData((String) finalData);
+                        } else {
+                            self.onData((byte[]) finalData);
                         }
-                    });
-                }
-            };
-            if (this.sslContext != null) {
-                this.ws.setWebSocketFactory(new DefaultSSLWebSocketClientFactory(this.sslContext));
+                    }
+                });
             }
-            this.ws.connect();
-        } catch (URISyntaxException e) {
-            throw new RuntimeException(e);
-        }
+
+            @Override
+            public void onPong(final Buffer payload) {
+                logger.fine("onPong");
+                EventThread.exec(new Runnable() {
+                    @Override
+                    public void run() {
+                        self.onPong();
+                    }
+                });
+            }
```

Comment:
I think we don't need this, because engine.io doesn't take care about ping/pong of websocket. Instead engine.io has a independent heartbeat mechanism.